tests/test_save.py

```diff
@@ -0,0 +1,250 @@
+    agent = ModularAgent(feature_set, function_set,
+                         when_learner=when,
+                         where_learner=where,
+                         planner='numbert',
+                         heuristic_learner=which,
+                         explanation_choice=choice,
+                         search_depth=2)
+    resp = req_A(agent)
+    assert resp == {}
+    
+    resp, ans = train_A_req_A(agent)
+    assert resp['selection'] == ans['selection']
+    assert resp['action'] == ans['action']
+    assert str(int(resp['inputs']['value'])) == ans['inputs']['value']
+
+    #check that copy is the same
+    dc = copy.deepcopy(resp)
+    assert(dc == resp)
+
+    #check that deserialized version is the same
+    agent_stream = dill.dumps(agent)
+    agent_deserialized = dill.loads(agent_stream)
+    assert(areAgentsSame(agent, agent_deserialized))
+
+    #make another request
+    resp = req_A(agent)
+    resp2 = req_A(agent_deserialized)
+    assert(resp == resp2)
+
+    #train the agents again
+    resp, ans = train_A_req_A(agent)
+    resp2, ans2 = train_A_req_A(agent_deserialized)
+
+    #where the state would be wrong
+    wrongAnswerState = {"arg1": {"above": "", "below": "hintWindow", "contentEditable": False, "dom_class": "CTATTextInput", "id": "arg1", "to_left": "", "to_right": "operator", "type": "TextField", "value": "2"},
+               "arg2": {"above": "", "below": "hintWindow", "contentEditable": False, "dom_class": "CTATTextInput", "id": "arg2", "to_left": "operator", "to_right": "answer", "type": "TextField", "value": "3"},
+               "answer": {"above": "", "below": "hint", "contentEditable": True, "dom_class": "CTATTextInput", "id": "answer", "to_left": "arg2", "to_right": "", "type": "TextField", "value": ""},
+               "hintWindow": {"above": "operator", "below": "", "dom_class": "CTATHintWindow", "id": "hintWindow", "to_left": "", "to_right": "hint", "type": "Component"},
+               "done": {"above": "hint", "below": "", "dom_class": "CTATDoneButton", "id": "done", "to_left": "hintWindow", "to_right": "", "type": "Component"},
+               "hint": {"above": "answer", "below": "done", "dom_class": "CTATHintButton", "id": "hint", "to_left": "hintWindow", "to_right": "", "type": "Component"},
+               "operator": {"above": "", "below": "hintWindow", "contentEditable": False, "dom_class": "CTATTextInput", "id": "operator", "to_left": "arg1", "to_right": "arg2", "type": "TextField", "value": "+"}}
+
+    #request with the wrongAnswerState and check sai
+    resp = agent.request(wrongAnswerState)
+    resp2 = agent_deserialized.request(wrongAnswerState)
+    assert(resp == resp2)
+
+
+    #check that agent does not get reinstantiated when deserialized
+    agent.__dict__['epsilon'] = 100
+    assert(agent.__dict__['epsilon'] == 100)
+    agent_stream = dill.dumps(agent)
+    agent_deserialized = dill.loads(agent_stream)
+    assert(agent_deserialized.__dict__['epsilon'] == 100)
```

Comment:
Not sure we need these final checks, since they are mainly checking against a possible language feature that doesn't exist.